components/scream/src/dynamics/homme/physics_dynamics_remapper.hpp

```diff
@@ -238,167 +293,467 @@ do_bind_field (const int ifield, const field_type& src, const field_type& tgt)
+
+      // Store dimensions of phys/dyn view
+      compute_view_dims<small_pack_type>(phys_alloc_prop, phys_dim, h_phys_dims(i));
+      compute_view_dims<small_pack_type>(dyn_alloc_prop,  dyn_dim,  h_dyn_dims(i));
+    } else {
+      h_pack_alloc_property(i) = AllocPropType::RealAlloc;
 
-    const bool phys_pack_alloc = phys_alloc_prop.template is_compatible<pack_type>();
-    const bool dyn_pack_alloc  = dyn_alloc_prop.template  is_compatible<pack_type>();
-    const bool phys_small_pack_alloc = phys_alloc_prop.template is_compatible<small_pack_type>();
-    const bool dyn_small_pack_alloc  = dyn_alloc_prop.template  is_compatible<small_pack_type>();
+      // Store dimensions of phys/dyn view
+      compute_view_dims<Real>(phys_alloc_prop, phys_dim, h_phys_dims(i));
+      compute_view_dims<Real>(dyn_alloc_prop,  dyn_dim,  h_dyn_dims(i));
+    }
 
-    const auto phys_lt = get_layout_type(ph.get_identifier().get_layout().tags());
+    // Store time levels
+    h_is_state_field_dev(i) = m_is_state_field[i];
+    h_time_levels(0).first  = tl.n0;
+    h_time_levels(0).second = tl.np1;
+  }
+
+  Kokkos::deep_copy(phys_ptrs,   h_phys_ptrs);
+  Kokkos::deep_copy(phys_layout, h_phys_layout);
+  Kokkos::deep_copy(phys_dims,   h_phys_dims);
+
+  Kokkos::deep_copy(dyn_ptrs,   h_dyn_ptrs);
+  Kokkos::deep_copy(dyn_layout, h_dyn_layout);
+  Kokkos::deep_copy(dyn_dims,   h_dyn_dims);
 
-    // phys->dyn requires a halo-exchange. Since not all entries in dyn
-    // are overwritten before the exchange, to avoid leftover garbage,
-    // we need to set all entries of dyn to zero.
-    if (m_is_state_field[i]) {
-      // Fill only the slice we need
-      if (phys_lt==LayoutType::Scalar2D) {
-        auto v = ekat::subview_1(dyn.template get_reshaped_view<Real****>(),itl);
-        Kokkos::deep_copy(v,0.0);
-      } else if (phys_lt==LayoutType::Scalar3D) {
-        auto v = ekat::subview_1(dyn.template get_reshaped_view<Real*****>(),itl);
-        Kokkos::deep_copy(v,0.0);
-      } else if (phys_lt==LayoutType::Vector3D) {
-        auto v = ekat::subview_1(dyn.template get_reshaped_view<Real******>(),itl);
-        Kokkos::deep_copy(v,0.0);
+  Kokkos::deep_copy(has_parent,          h_has_parent);
+  Kokkos::deep_copy(time_levels,         h_time_levels);
+  Kokkos::deep_copy(pack_alloc_property, h_pack_alloc_property);
+  Kokkos::deep_copy(is_state_field_dev,  h_is_state_field_dev);
+}
+
+template<typename RealType>
+template <typename ScalarT, typename MT>
+KOKKOS_FUNCTION
+void PhysicsDynamicsRemapper<RealType>::
+set_dyn_to_zero(const MT& team) const
+{
+  const int i = team.league_rank();
+  const int itl = time_levels(0).first;
+  const auto& dim_d = dyn_dims(i).dims;
+
+  switch (phys_layout(i)) {
+    case etoi(LayoutType::Scalar2D):
+    {
+      if (is_state_field_dev(i)) {
+        auto dyn = Unmanaged<KokkosTypes<DefaultDevice>::view<ScalarT****>> (reinterpret_cast<ScalarT*>(dyn_ptrs(i).get()),
```

Comment:
These lines are quite long to parse. You might want to have in the class, something like

```
using device_type = typename field_type::device_type;
template<typename T, int N>
using uview_nd = Unmanaged<KokkosTypes<device_type>::view_ND<T,N>>;
```
And then, in the different fcns where you need them, simply do
```
auto dyn = uview_4d<ScalarT>(...);
```
It might be easier to parse.

Reply:
And cleaner looking..